File: src/main/java/com/example/demo/config/SecurityConfig.java
package com.example.demo.config;

import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import javax.sql.DataSource;

import org.springframework.beans.factory.ObjectProvider;
import org.springframework.boot.web.servlet.ServletListenerRegistrationBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.HeadersConfigurer;
import org.springframework.security.config.annotation.web.configurers.HttpBasicConfigurer;
import org.springframework.security.config.annotation.web.configurers.SessionManagementConfigurer;
import org.springframework.security.core.session.SessionRegistry;
import org.springframework.security.core.session.SessionRegistryImpl;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.provisioning.JdbcUserDetailsManager;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.security.web.authentication.LoginUrlAuthenticationEntryPoint;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.authentication.session.CompositeSessionAuthenticationStrategy;
import org.springframework.security.web.authentication.session.ConcurrentSessionControlAuthenticationStrategy;
import org.springframework.security.web.authentication.session.RegisterSessionAuthenticationStrategy;
import org.springframework.security.web.authentication.session.SessionAuthenticationStrategy;
import org.springframework.security.web.authentication.session.SessionFixationProtectionStrategy;
import org.springframework.security.web.csrf.CookieCsrfTokenRepository;
import org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler;
import org.springframework.security.web.header.writers.ReferrerPolicyHeaderWriter;
import org.springframework.security.web.header.writers.StaticHeadersWriter;
import org.springframework.security.web.header.writers.XXssProtectionHeaderWriter;
import org.springframework.security.web.session.HttpSessionEventPublisher;
import org.springframework.session.FindByIndexNameSessionRepository;
import org.springframework.session.Session;
import org.springframework.session.security.SpringSessionBackedSessionRegistry;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import lombok.RequiredArgsConstructor;

@Configuration
@EnableWebSecurity
@EnableMethodSecurity
@RequiredArgsConstructor
public class SecurityConfig {

        private static final long HSTS_MAX_AGE = 31536000L;
        private static final long CORS_MAX_AGE = 3600L;
        private static final List<String> ALLOWED_ORIGINS = List.of(
                        "https://localhost:8443",
                        "http://localhost:8080",
                        "https://infoapp.org");
        private static final List<String> ALLOWED_METHODS = List.of("GET", "POST", "PUT", "DELETE", "OPTIONS");

        private final AuthenticationFailureHandler customAuthenticationFailureHandler;
        private final LoginRateLimitFilter loginRateLimitFilter;

        private final AppProperties appProperties;
        private final WebAuthnProperties webAuthnProperties;

        @Bean
        public SecurityFilterChain filterChain(HttpSecurity http, SessionRegistry sessionRegistry) {
                String cspConnectSrc = appProperties.getCsp().getConnectSrc();
                String cspPolicy = "default-src 'self'; " +
                                "script-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; " +
                                "style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; " +
                                "img-src 'self' data: https:; " +
                                "font-src 'self' data: https://cdn.jsdelivr.net; " +
                                "connect-src 'self' https://date.nager.at " + cspConnectSrc + "; " +
                                "frame-ancestors 'none'; " +
                                "base-uri 'self'; " +
                                "form-action 'self'";

                http
                                .csrf(csrf -> {
                                        CookieCsrfTokenRepository csrfTokenRepository = CookieCsrfTokenRepository
                                                        .withHttpOnlyFalse();
                                        XorCsrfTokenRequestAttributeHandler requestHandler = new XorCsrfTokenRequestAttributeHandler();
                                        csrf.csrfTokenRepository(csrfTokenRepository);
                                        csrf.csrfTokenRequestHandler(requestHandler);
                                        csrf.ignoringRequestMatchers("/api/webauthn/**");
                                })
                                .cors(cors -> cors.configurationSource(corsConfigurationSource()))
                                .headers(headers -> headers
                                                .frameOptions(HeadersConfigurer.FrameOptionsConfig::deny)
                                                .contentTypeOptions(Customizer.withDefaults())
                                                .xssProtection(xss -> xss.headerValue(
                                                                XXssProtectionHeaderWriter.HeaderValue.ENABLED_MODE_BLOCK))
                                                .httpStrictTransportSecurity(hsts -> hsts.includeSubDomains(true)
                                                                .maxAgeInSeconds(HSTS_MAX_AGE))
                                                .referrerPolicy(referrer -> referrer.policy(
                                                                ReferrerPolicyHeaderWriter.ReferrerPolicy.STRICT_ORIGIN_WHEN_CROSS_ORIGIN))
                                                .contentSecurityPolicy(csp -> csp.policyDirectives(cspPolicy))
                                                .addHeaderWriter(new StaticHeadersWriter(
                                                                "Permissions-Policy",
                                                                "geolocation=(), microphone=(), camera=(), usb=(), payment=(), fullscreen=()"))
                                                .cacheControl(Customizer.withDefaults()))
                                .authorizeHttpRequests(auth -> auth
                                                .requestMatchers(
                                                                "/css/**", "/js/**", "/webjars/**",
                                                                "/favicon.ico", "/favicon.svg", "/.well-known/**",
                                                                "/login", "/api/webauthn/**")
                                                .permitAll()
                                                .requestMatchers("/tools/passkey", "/tools/passkey/**").authenticated()
                                                .requestMatchers("/", "/home", "/home/**", "/tools", "/tools/**")
                                                .permitAll()
                                                .requestMatchers("/timesheet", "/timesheet/**").hasRole("UNISS")
                                                .requestMatchers("/api/calendar/holidays").permitAll()
                                                .requestMatchers("/api/calendar/holidays/list").permitAll()
                                                .requestMatchers("/api/calendar/holidays/range").permitAll()
                                                .requestMatchers("/api/calendar/holidays/**").authenticated()
                                                .requestMatchers("/manage/**").hasRole("ADMIN")
                                                .requestMatchers("/api/**").authenticated()
                                                .requestMatchers("/actuator/health", "/actuator/info").permitAll()
                                                .requestMatchers("/actuator/**").denyAll()
                                                .requestMatchers("/error").permitAll()
                                                .anyRequest().authenticated())
                                .exceptionHandling(ex -> ex
                                                .defaultAuthenticationEntryPointFor(
                                                                new LoginUrlAuthenticationEntryPoint("/tools"),
                                                                request -> request.getRequestURI() != null
                                                                                && request.getRequestURI().startsWith(
                                                                                                "/tools/passkey")))
                                .addFilterBefore(loginRateLimitFilter, UsernamePasswordAuthenticationFilter.class)
                                .formLogin(form -> form
                                                .loginPage("/login").permitAll()
                                                .defaultSuccessUrl("/tools")
                                                .failureHandler(customAuthenticationFailureHandler))
                                .webAuthn(webAuthn -> {
                                        webAuthn.rpName(webAuthnProperties.getName());
                                        webAuthn.rpId(webAuthnProperties.getId());
                                        Set<String> origins;
                                        if (webAuthnProperties.getAllowedOrigins() != null && !webAuthnProperties.getAllowedOrigins().isBlank()) {
                                                origins = Arrays.stream(webAuthnProperties.getAllowedOrigins().split(","))
                                                                .map(String::trim)
                                                                .collect(Collectors.toSet());
                                        } else {
                                                origins = Set.of(webAuthnProperties.getOrigin());
                                        }
                                        webAuthn.allowedOrigins(origins);
                                })
                                .httpBasic(HttpBasicConfigurer::disable)
                                .logout(logout -> logout
                                                .logoutUrl("/logout")
                                                .logoutSuccessUrl("/tools")
                                                .invalidateHttpSession(true)
                                                .deleteCookies("JSESSIONID", "SESSION")
                                                .clearAuthentication(true)
                                                .permitAll())
                                .sessionManagement(sess -> sess
                                                .sessionFixation(
                                                                SessionManagementConfigurer.SessionFixationConfigurer::migrateSession)
                                                .invalidSessionUrl("/tools")
                                                .maximumSessions(1)
                                                .maxSessionsPreventsLogin(false)
                                                .expiredUrl("/tools")
                                                .sessionRegistry(sessionRegistry));

                return http.build();
        }

        /**
         * CORS設定: 本番環境では適切に制限すること
         */
        @Bean
        public CorsConfigurationSource corsConfigurationSource() {
                CorsConfiguration configuration = new CorsConfiguration();
                String allowed = System.getenv("ALLOWED_ORIGINS");
                configuration.setAllowedOrigins(allowed != null ? Arrays.asList(allowed.split(",")) : ALLOWED_ORIGINS);
                configuration.setAllowedMethods(ALLOWED_METHODS);
                configuration.setAllowedHeaders(List.of("*"));
                configuration.setAllowCredentials(true);
                configuration.setMaxAge(CORS_MAX_AGE);

                UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
                source.registerCorsConfiguration("/**", configuration);
                return source;
        }

        @Bean
        public PasswordEncoder passwordEncoder() {
                return PasswordEncoderFactories.createDelegatingPasswordEncoder();
        }

        /**
         * ユーザー認証（DB: users / authorities テーブル、Flyway V8 でスキーマ・初期データ投入）
         * <p>
         * パスワード変更・ユーザー追加は DB を更新するか、JdbcUserDetailsManager の API を利用すること。
         */
        @Bean
        public UserDetailsService userDetailsService(DataSource dataSource) {
                return new JdbcUserDetailsManager(dataSource);
        }

        @Bean
        public SessionRegistry sessionRegistry(ObjectProvider<FindByIndexNameSessionRepository<Session>> repoProvider) {
                FindByIndexNameSessionRepository<Session> repo = repoProvider.getIfAvailable();
                if (repo != null) {
                        return new SpringSessionBackedSessionRegistry<>(repo);
                }
                return new SessionRegistryImpl();
        }

        @Bean
        public SessionAuthenticationStrategy sessionAuthenticationStrategy(SessionRegistry sessionRegistry) {
                ConcurrentSessionControlAuthenticationStrategy concurrent = new ConcurrentSessionControlAuthenticationStrategy(
                                sessionRegistry);
                concurrent.setMaximumSessions(1);
                concurrent.setExceptionIfMaximumExceeded(false);

                SessionFixationProtectionStrategy fixation = new SessionFixationProtectionStrategy();
                RegisterSessionAuthenticationStrategy register = new RegisterSessionAuthenticationStrategy(
                                sessionRegistry);
                return new CompositeSessionAuthenticationStrategy(List.of(concurrent, fixation, register));
        }

        @Bean
        public ServletListenerRegistrationBean<HttpSessionEventPublisher> httpSessionEventPublisher() {
                return new ServletListenerRegistrationBean<>(new HttpSessionEventPublisher());
        }
}

File: src/main/kotlin/com/example/demo/config/SecurityConfig.kt
package com.example.demo.config

import org.springframework.beans.factory.ObjectProvider
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.web.servlet.ServletListenerRegistrationBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.core.session.SessionRegistry
import org.springframework.security.core.session.SessionRegistryImpl
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.provisioning.InMemoryUserDetailsManager
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.csrf.CookieCsrfTokenRepository
import org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler
import org.springframework.security.web.header.writers.ReferrerPolicyHeaderWriter
import org.springframework.security.web.header.writers.StaticHeadersWriter
import org.springframework.security.web.header.writers.XXssProtectionHeaderWriter
import org.springframework.security.web.session.HttpSessionEventPublisher
import org.springframework.session.FindByIndexNameSessionRepository
import org.springframework.session.Session
import org.springframework.session.security.SpringSessionBackedSessionRegistry
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)
class SecurityConfig(
    private val customAuthenticationFailureHandler: AuthenticationFailureHandler,
    private val loginRateLimitFilter: LoginRateLimitFilter
) {

    @Value("\${app.csp.connect-src:'self'}")
    private lateinit var cspConnectSrc: String

    @Value("\${webauthn.rp.id:localhost}")
    private lateinit var webAuthnRpId: String

    @Value("\${webauthn.rp.name:Dev RP}")
    private lateinit var webAuthnRpName: String

    @Value("\${webauthn.rp.origin:http://localhost:8080}")
    private lateinit var webAuthnRpOrigin: String

    @Value("\${webauthn.rp.allowed-origins:}")
    private var webAuthnAllowedOrigins: String = ""

    companion object {
        private const val HSTS_MAX_AGE = 31536000L // 1年
        private const val CORS_MAX_AGE = 3600L // 1時間
        private val ALLOWED_ORIGINS = listOf("https://localhost:8443", "http://localhost:8080", "https://infoapp.org")
        private val ALLOWED_METHODS = listOf("GET", "POST", "PUT", "DELETE", "OPTIONS")
    }

    @Bean
    fun filterChain(http: HttpSecurity, sessionRegistry: SessionRegistry): SecurityFilterChain {
        val cspPolicy = buildString {
            append("default-src 'self'; ")
            append("script-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; ")
            append("style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; ")
            append("img-src 'self' data: https:; ")
            append("font-src 'self' data: https://cdn.jsdelivr.net; ")
            append("connect-src 'self' https://date.nager.at ").append(cspConnectSrc).append("; ")
            append("frame-ancestors 'none'; ")
            append("base-uri 'self'; ")
            append("form-action 'self'")
        }

        http
            // CSRF保護を有効化（CookieベースのCSRFトークン + BREACH攻撃対策）
            .csrf { csrf ->
                val csrfTokenRepository = CookieCsrfTokenRepository.withHttpOnlyFalse()
                val requestHandler = XorCsrfTokenRequestAttributeHandler()
                requestHandler.setCsrfRequestAttributeName(null) // リクエスト属性名をnullにしてdeferred token loadを有効化
                csrf.csrfTokenRepository(csrfTokenRepository)
                csrf.csrfTokenRequestHandler(requestHandler)
                csrf.ignoringRequestMatchers("/api/webauthn/**") // WebAuthnはフロントJSから直接叩くのでCSRF除外
            }

            // CORS設定
            .cors { it.configurationSource(corsConfigurationSource()) }

            // セキュリティヘッダーの設定
            .headers { headers ->
                headers
                    // X-Frame-Options: DENY（クリックジャッキング対策）
                    .frameOptions { it.deny() }
                    // X-Content-Type-Options: nosniff
                    .contentTypeOptions { }
                    // X-XSS-Protection: 1; mode=block
                    .xssProtection { it.headerValue(XXssProtectionHeaderWriter.HeaderValue.ENABLED_MODE_BLOCK) }
                    // Strict-Transport-Security（HTTPS使用時に有効化）
                    .httpStrictTransportSecurity { it.includeSubDomains(true).maxAgeInSeconds(HSTS_MAX_AGE) }
                    // Referrer-Policy: 外部サイトへのリファラー情報を制限
                    .referrerPolicy { it.policy(ReferrerPolicyHeaderWriter.ReferrerPolicy.STRICT_ORIGIN_WHEN_CROSS_ORIGIN) }
                    // Content-Security-Policy: XSS攻撃対策
                    .contentSecurityPolicy { it.policyDirectives(cspPolicy) }
                    .addHeaderWriter(
                        StaticHeadersWriter(
                            "Permissions-Policy",
                            "geolocation=(), microphone=(), camera=(), usb=(), payment=(), fullscreen()"
                        )
                    )
                    // キャッシュ制御: ブラウザの戻るボタンでキャッシュページを表示させない
                    .cacheControl { }
            }

            // 認可設定
            .authorizeHttpRequests { auth ->
                auth
                    // 静的リソースは認証不要
                    .requestMatchers(
                        "/css/**", "/js/**", "/webjars/**",
                        "/favicon.ico", "/favicon.svg", "/.well-known/**",
                        "/login", "/api/webauthn/**"
                    ).permitAll()
                    // メイン画面は認証不要（閲覧のみ）
                    .requestMatchers("/", "/home", "/home/**", "/tools", "/tools/**").permitAll()
                    // 勤務表はUNISSロールのみアクセス可能
                    .requestMatchers("/timesheet", "/timesheet/**").hasRole("UNISS")
                    // 祝日APIは閲覧のみ認証不要
                    .requestMatchers("/api/calendar/holidays").permitAll()
                    .requestMatchers("/api/calendar/holidays/list").permitAll()
                    .requestMatchers("/api/calendar/holidays/range").permitAll()
                    .requestMatchers("/api/calendar/holidays/**").authenticated() // 追加・削除は認証必須
                    // 管理画面は認証必須
                    .requestMatchers("/manage/**").hasRole("ADMIN")
                    // その他のAPIは認証必須
                    .requestMatchers("/api/**").authenticated()
                    // Actuatorエンドポイントは制限
                    .requestMatchers("/actuator/health", "/actuator/info").permitAll()
                    .requestMatchers("/actuator/**").denyAll()
                    // エラーページは許可
                    .requestMatchers("/error").permitAll()
                    // その他は認証必須（デフォルト拒否の原則）
                    .anyRequest().authenticated()
            }
            // ログイン前にログイン試行のレート制限フィルターを追加
            .addFilterBefore(loginRateLimitFilter, UsernamePasswordAuthenticationFilter::class.java)

            // フォームログインを有効化
            .formLogin {
                it.loginPage("/login").permitAll()
                it.defaultSuccessUrl("/tools")
                it.failureHandler(customAuthenticationFailureHandler)
            }
            .webAuthn {
                it.rpName(webAuthnRpName)
                it.rpId(webAuthnRpId)
                // 複数オリジンをサポート（カンマ区切り）
                val origins = if (webAuthnAllowedOrigins.isNotBlank()) {
                    webAuthnAllowedOrigins.split(",").map { o -> o.trim() }.toSet()
                } else {
                    setOf(webAuthnRpOrigin)
                }
                it.allowedOrigins(origins)
            }
            .httpBasic { it.disable() }
            .logout {
                it.logoutUrl("/logout")
                it.logoutSuccessUrl("/login?logout=true") // ログアウト後はログインモーダル表示
                it.invalidateHttpSession(true) // セッションを無効化
                it.deleteCookies("JSESSIONID", "SESSION") // セッションCookieを削除
                it.clearAuthentication(true) // 認証情報をクリア
                it.permitAll()
            }

            // セッション管理: セッション無効（タイムアウトや強制ログアウト）時のリダイレクト先
            // および同一ユーザ最大セッション数を 1 にして古いセッションを切断（新ログインを許可）
            .sessionManagement { sess ->
                sess.sessionFixation { it.migrateSession() } // セッション固定攻撃対策
                sess.invalidSessionUrl("/tools") // セッション無効時もツールへ
                sess.maximumSessions(1)
                    .maxSessionsPreventsLogin(false)
                    .expiredUrl("/tools") // セッション期限切れもツールへ
                    .sessionRegistry(sessionRegistry)
            }

        return http.build()
    }

    /**
     * CORS設定: 本番環境では適切に制限すること
     */
    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration().apply {
            // 本番環境では特定のオリジンのみ許可すべき
            allowedOrigins = System.getenv("ALLOWED_ORIGINS")?.split(",") ?: ALLOWED_ORIGINS
            allowedMethods = ALLOWED_METHODS
            allowedHeaders = listOf("*")
            allowCredentials = true
            maxAge = CORS_MAX_AGE
        }
        return UrlBasedCorsConfigurationSource().apply {
            registerCorsConfiguration("/**", configuration)
        }
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder = PasswordEncoderFactories.createDelegatingPasswordEncoder()

    /**
     * ユーザー認証設定
     *
     * 【本番環境での注意事項】
     * - InMemoryUserDetailsManagerは開発用。本番ではDBまたはLDAP/OIDC等を使用すること
     * - パスワードは最低12文字以上、大文字・小文字・数字・記号を含むこと
     * - ユーザー名とパスワードを同一にしないこと
     */
    @Bean
    fun userDetailsService(passwordEncoder: PasswordEncoder): UserDetailsService {
        val user1 = User.builder()
            .username("角谷亮洋")
            .password(passwordEncoder.encode("角谷亮洋"))
            .roles("USER", "UNISS") // UNISSロールを追加（勤務表アクセス用）
            .build()

        // 管理者ユーザー（ADMINロール）
        val admin = User.builder()
            .username("admin")
            .password(passwordEncoder.encode("Admin@2026!Secure"))
            .roles("ADMIN")
            .build()

        return InMemoryUserDetailsManager(user1, admin)
    }

    @Bean
    fun sessionRegistry(repoProvider: ObjectProvider<FindByIndexNameSessionRepository<Session>>): SessionRegistry {
        val repo = repoProvider.ifAvailable
        return if (repo != null) {
            SpringSessionBackedSessionRegistry(repo)
        } else {
            SessionRegistryImpl()
        }
    }

    @Bean
    fun httpSessionEventPublisher(): ServletListenerRegistrationBean<HttpSessionEventPublisher> {
        return ServletListenerRegistrationBean(HttpSessionEventPublisher())
    }
}
